backend/internal/service/oidc_service.go

```diff
@@ -277,6 +278,93 @@ func (s *OidcService) createTokenFromRefreshToken(refreshToken, clientID, client
 	return accessToken, newRefreshToken, 3600, nil
 }
 
+func (s *OidcService) IntrospectToken(clientID, clientSecret, tokenString string) (introspectDto dto.OidcIntrospectionResponseDto, err error) {
+	if clientID == "" || clientSecret == "" {
+		return introspectDto, &common.OidcMissingClientCredentialsError{}
+	}
+
+	// Get the client to check if we are authorized.
+	var client model.OidcClient
+	if err := s.db.First(&client, "id = ?", clientID).Error; err != nil {
+		return introspectDto, &common.OidcClientSecretInvalidError{}
+	}
+
+	// Verify the client secret. This endpoint may not be used by public clients.
+	if client.IsPublic {
+		return introspectDto, &common.OidcClientSecretInvalidError{}
+	}
+
+	if err := bcrypt.CompareHashAndPassword([]byte(client.Secret), []byte(clientSecret)); err != nil {
+		return introspectDto, &common.OidcClientSecretInvalidError{}
+	}
+
+	token, err := s.jwtService.VerifyIdToken(tokenString, false)
+	if err != nil {
+		if errors.Is(err, jwt.ParseError()) {
+			// It's apparently not a valid JWT token, so we check if it's a valid refresh_token.
+			return s.introspectRefreshToken(tokenString)
```

Comment:
I think you could do this in a more heuristic way.

Refresh tokens issues by Pocket ID are fixed at 40-chars and have no dots. So before attempting to parse this as access token, you could simply check if the length is 40 chars and has no dots (`!strings.ContainsRune`) and treat it as refresh token.

Also note the specs have a `token_hint` parameter which can be used by clients to hint if it's an access token or refresh token (not required)

Reply:
True, however the other endpoints that handle refreshTokens also don't apply any checks besides the DB lookup. The only scenario I see where this could hurt us is, if someone deliberately fires random stuff against the `introspect` endpoint. If they wanted to do that, they could as well throw that random stuff against the `token` endpoint, though.

So basically: I opted for keeping the logic unified and didn't implement it differently than in other spots.